### PR TITLE
Add `serialized_len` method in `Encodable` trait

### PR DESCRIPTION
**Pull request by RCasatta**
It's nice to know ahead of time how much space an encodable object will take once serialized, this allows to allocate the capacity of a vector before using it for storing serialized information, avoiding vector re-allocations.

Used in the `serialize` method here https://github.com/rust-bitcoin/rust-bitcoin/blob/99ae48ab010fa6773687ca96deff46ab63fadc12/src/consensus/encode.rs#L142 as done in the second commit, improve serialize performance a lot for object sized hundreds of bytes where the weight of re-allocations in comparison of the serialization work is bigger. 

```
# before                                                                        
                                                                                
test blockdata::transaction::benches::bench_transaction_serialize               ... bench:         550 ns/iter (+/- 31)
                                                                                
# after                                                                         
                                                                                
test blockdata::transaction::benches::bench_transaction_serialize               ... bench:         144 ns/iter (+/- 4)
```
(Note I renamed current benchmark to `bench_transaction_serialize_with_capacity` because it was using a pre-created workhorse collection to store serialization data and created a new one `bench_transaction_serialize` which use `serialize()`)

Still a draft cause I have to re-check everything but I will do after https://github.com/rust-bitcoin/rust-bitcoin/pull/988 is merged so that I don't have to bother to implement `serialized_len` in psbt methods .

### CODE REVIEW

**Comment by dpc** [comment]
We should be mindful about definition of a contract here. 

* Are we always commit to providing fast yet 100% accurate size? Could users e.g. depend on it writing `unsafe` code?
* Would be be better to define it as something less precise?
  * `fn estimated_serialized_len`
  * `fn maximum_serialized_len` (serialized data guaranteed to be less or equal)

I could imagine that `maximum_serialized_len` could be marginally faster (no need to inspect e.g. varints), at the cost of over-allocating a little bit, etc.

File: src/consensus/encode.rs [hunk 1]
```diff
@@ -139,9 +139,11 @@ impl From<psbt::Error> for Error {
 
 /// Encode an object into a vector
 pub fn serialize<T: Encodable + ?Sized>(data: &T) -> Vec<u8> {
-    let mut encoder = Vec::new();
+    let serialized_len = data.serialized_len();
+    let mut encoder = Vec::with_capacity(serialized_len);
     let len = data.consensus_encode(&mut encoder).expect("in-memory writers don't error");
     debug_assert_eq!(len, encoder.len());
+    debug_assert_eq!(serialized_len, encoder.len());
```
Comment (tcharding) [inline src/consensus/encode.rs:146]: This implicitly unit tests the implementations of `serialize_len()` as long as we serialize somewhere in unit tests, right? I had a look at `consensus/encode.rs` and there are a bunch of tests but I don't know if we have full coverage. Did you do this on purpose? If so, perhaps when the final version goes up could you add some text justifying not adding explicit unit tests for `serialize_len` please so other reviewers don't wonder why there is none added?

**Review by tcharding — COMMENTED** [review]
LGTM, since its draft just mentioned the unit tests and one other thing, it was not immediately clear what all the magic numbers meant, perhaps in the final version we could have comments for them (assuming its not just unclear in the diff and is actually obvious in the code). Thanks.

**Comment by RCasatta** [comment]
> We should be mindful about definition of a contract here.
> 
>     * Can we always commit to providing fast yet 100% accurate size? Could users e.g. depend on it writing `unsafe` code?
> 
>     * Would it make sense to define something less precise instead/as well?
>       
>       * `fn estimated_serialized_len`
>       * `fn maximum_serialized_len` (serialized data guaranteed to be less or equal)
> 
> 
> I could imagine that `maximum_serialized_len` could be marginally faster (no need to inspect e.g. varints), at the cost of over-allocating a little bit, etc. Or maybe precise is always fast and easy. Just asking. :D

I think being accurate it's not an issue, the implicit `debug_assert` in `serialize` already gives me some confidence (and as suggested by @tcharding I will also add some explicit tests in the final version).

While being fast, you are right, it's an issue for big objects like blocks where this technique is not performing well. The reason is that you have fewer reallocations the bigger the objects become. However, gains for small objects are significant and there are a lot of downstream projects serializing a tx or its parts (Example [1](https://github.com/RCasatta/blocks_iterator/blob/7ed3c8cd440bd39c37245bdb304a3c05536fcd23/src/utxo/db.rs#L133), [2](https://github.com/bitcoindevkit/bdk/blob/fbd98b4c5a88a10f8dfc22b4c54187eba4d68afe/src/database/memory.rs#L69))

My proposal is to add a parameter `cap: usize` to `serialized_len` and break the sum calculation in `impl Encodable for Vec<$type>` if the partial result is greater than `cap`. So the contract of the fn becomes that the result is exact if lower than `cap`, while if greater is just a min bound (obviously, calling with `usize::MAX` still gives exact result if needed)

Then inside `fn serialize` use something with a reasonable value like `Vec::with_capacity(self.serialized_len(1000))`

> LGTM, since its draft just mentioned the unit tests and one other thing, it was not immediately clear what all the magic numbers meant, perhaps in the final version we could have comments for them (assuming its not just unclear in the diff and is actually obvious in the code). Thanks.

I will add more comments in the final version

File: src/consensus/encode.rs (hunk 1 above)
Comment (RCasatta) [reply src/consensus/encode.rs:146]: > This implicitly unit tests the implementations of serialize_len() as long as we serialize somewhere in unit tests, right?

Yes, that was the idea to have a quick check, I will add the comment and also some explicit unit tests in the final version.

**Comment by Kixunil** [comment]
This is awesome!

> Could users e.g. depend on it writing `unsafe` code?

No and this is clearly communicated by the *trait* **not** being `unsafe`. However I wouldn't mind spelling it out in docs for newbies. Also note that the current policy of no `unsafe` prevents adding `unsafe` to the trait.

> My proposal is to add a parameter `cap: usize` to `serialized_len` and break the sum calculation

I don't understand how this could lead to speed improvement, I guess it'll be worse. But I'm not sure I understand what you describe correctly. Obviously benchmarks will be needed to check this anyway.

**Comment by RCasatta** [comment]
> I don't understand how this could lead to speed improvement, I guess it'll be worse. But I'm not sure I understand what you describe correctly. Obviously benchmarks will be needed to check this anyway.

Didn't implement the `cap` thing yet, but have a look at bench result with this MR plus this commit https://github.com/rust-bitcoin/rust-bitcoin/commit/5705e88685044f2301c0f104f43c4963ec8554dd

```
test blockdata::block::benches::bench_block_serialize                           ... bench:     942,927 ns/iter (+/- 16,702)
test blockdata::block::benches::bench_block_serialize_realloc                   ... bench:     837,625 ns/iter (+/- 19,035)
test blockdata::block::benches::bench_block_serialize_with_capacity             ... bench:     833,153 ns/iter (+/- 6,677)
test blockdata::block::benches::bench_block_serialized_len                      ... bench:      82,208 ns/iter (+/- 1,974)
test blockdata::transaction::benches::bench_transaction_serialize               ... bench:         145 ns/iter (+/- 5)
test blockdata::transaction::benches::bench_transaction_serialize_realloc       ... bench:         571 ns/iter (+/- 26)
test blockdata::transaction::benches::bench_transaction_serialize_with_capacity ... bench:         108 ns/iter (+/- 8)
test blockdata::transaction::benches::bench_transaction_serialized_len          ... bench:          10 ns/iter (+/- 0)

```
As you can see, performance of block serialization with serialized_len estimation is worse because the cost of computing the len is non-negligible.
The change is so effective for small objects and bad for big one because for an object of 1k you do 10 re-allocations, while for an object of 1M you do just 20 re-allocations (vector capacity doubles). While computing the serialized_len grows linearly with the size of the object

**Comment by RCasatta** [comment]
pushed the `serialized_len(threshold: usize)` version.

Still very rough code, commits to be reorganized, but benchmarks are ok:

```
test blockdata::block::benches::bench_block_serialize                           ... bench:     837,149 ns/iter (+/- 6,638)
test blockdata::block::benches::bench_block_serialize_realloc                   ... bench:     840,989 ns/iter (+/- 5,990)
test blockdata::block::benches::bench_block_serialize_with_capacity             ... bench:     835,592 ns/iter (+/- 5,488)
test blockdata::block::benches::bench_block_serialized_len                      ... bench:          63 ns/iter (+/- 3)

test blockdata::transaction::benches::bench_transaction_serialize               ... bench:         164 ns/iter (+/- 9)
test blockdata::transaction::benches::bench_transaction_serialize_realloc       ... bench:         547 ns/iter (+/- 21)
test blockdata::transaction::benches::bench_transaction_serialize_with_capacity ... bench:         107 ns/iter (+/- 3)
test blockdata::transaction::benches::bench_transaction_serialized_len          ... bench:          12 ns/iter (+/- 0)
```

**Comment by Kixunil** [comment]
Ah, I think I understand now.

"The method is intended for allocation optimizations and should return a precise value if possible and fast however it may return a bit greater than the actual serialized length if the computation of the value becomes less expensive assuming sacrificing a bit of RAM bytes is worth it."

Would sound good to me. I can already imagine that ignoring the actual length of compact size and assuming largest case may perform better while wasting only a few % of bytes.

**Comment by dpc** [comment]
> Ah, I think I understand now.

I'm not sure what you're referring to exactly.

My understanding is that calculating the size of the block is the biggest issue. One mainnet block is > 2k txes, and it's impossible be precise without actually going through all the data. To lesser extent the same is recursively a problem for each `Tx` itself with inputs and outputs.

> it may return a bit greater


I think it's best to commit to nothing for the estimate, other than "doing our best". For smaller types returning precise size, for larger using whatever heuristic that works well with mainnet data. Eg. an estimated size of a block is "AVG_MAINNET_TX_SIZE * block.txs.len() + header size + 16 /* some extras just to avoid reallocation */", etc.

Since this is a pure heuristic, it will never be perfect, and there's multiple ways to approach it, but for real world performance giving a quick guess that doesn't cause buffer reallocation 95% of the time and doesn't waste too much memory, would be a huge improvement no matter what.

**Comment by RCasatta** [comment]
@Kixunil I see what you mean with `is_big_object` now, and I think it's cleaner than adding a parameter to `serialized_len` so I will change this MR accordingly

@dpc I think heuristics of serialized space makes sense but having an accurate `serialized_len` (explaining it could be slow in fn docs) together with `is_big_object` will accomplish my goal for this PR

**Comment by RCasatta** [comment]
Implemented `is_big_object` as suggested by @Kixunil 

PR still needs tests and additional comments on "magic numbers" returned by serialized_len, I will update after #988

**Comment by Kixunil** [comment]
@RCasatta cool! Did you try to check how various values for `max_iterations` affect the speed? Maybe it'd even make sense to have a helper for some specific good value if we can consistently reproduce it as the best one.

**Comment by RCasatta** [comment]
> Did you try to check how various values for max_iterations affect the speed? 
 
No, to start I chose a safe low value (10) for `max_iterations` where I am sure there is no performance degradation for big objects but at the same time the performance gains are covering most of the use cases.

**Comment by Kixunil** [comment]
I don't think it matters that you reallocate if the program is going to crash on OOM anyway.

**Comment by dpc** [comment]
@Kixunil What if it doesn't crash, but just severely degrades the performance?

**Comment by Kixunil** [comment]
Assuming otherwise correct implementation, if it overflows `usize` then the data would've been bigger than what fits into RAM anyway. If it overflows because it's wrong it could affect performance already for other reasons.

**Comment by dpc** [comment]
Oh, I see what you mean. Nevermind, my point is moot.

Though even in precise version - isn't serialized data always smaller than deserialized one? To overflow the `usize` user would have to call it with input that takes more memory than `usize` can handle, which IIRC is not possible.

Though I guess there is no guarantee that whatever implemented `Encodable` acts in a "normal" way.

OK, your point holds. :+1:

**Comment by Kixunil** [comment]
`Enodable` is a public, non-sealed trait so one could make (a buggy) transaction generator that generates super-large txes. Also I can think of one ridiculous example where serialized size is larger: compact size for values >= 2^32 (8B in memory, 9B serialized). :)

**Comment by RCasatta** [comment]
I was thinking, instead of `is_big_object()` to use a `serialized_len_min()` which returns the minimal serialized size of an `Encodable` object.

Even for blocks with many transactions is about as fast as `is_big_object()` (no need to loop, just a multiplication) and can be similarly used as a discriminant to decide to use the precise value or the minimal one in `serialize()`.
